core/node/l1_recovery/src/l1_fetcher/l1_fetcher.rs

```diff
@@ -0,0 +1,1083 @@
+    }
+
+    pub async fn find_block_near_genesis(&self) -> U64 {
+        let mut start_block: U64 = U64::zero();
+        let last_l1_block_number = L1Fetcher::get_last_l1_block_number(&self.eth_client)
+            .await
+            .unwrap();
+        let mut end_block = last_l1_block_number;
+        // we use step override as such big step value guarantees that we don't miss
+        // events on non-local networks
+        let step = 50_000;
+        while start_block + 2 * step < end_block {
+            let mid = (start_block + end_block) / 2;
+            let logs = self
+                .get_logs(mid, mid + step - 1, RollupEventType::Commit)
+                .await;
+            if !logs.is_empty() {
+                end_block = mid;
+            } else {
+                start_block = mid;
+            }
+        }
+        end_block = last_l1_block_number;
+        let mut current_block = start_block;
+        loop {
+            let filter_to_block = cmp::min(current_block + step - 1, end_block);
+            let logs = self
+                .get_logs(
+                    current_block,
+                    filter_to_block,
+                    RollupEventType::NewPriorityTx,
+                )
+                .await;
+            if !logs.is_empty() {
+                if L1Fetcher::extract_priority_op_id_from_rollup_event_log(
+                    RollupEventType::NewPriorityTx,
+                    &logs[0],
+                ) == PriorityOpId(0)
+                {
+                    return logs[0].block_number.unwrap();
+                } else {
+                    // The binary search didn't work, it's fine, we just return first block
+                    return U64::zero();
+                }
+            }
+            if filter_to_block == end_block {
+                // The binary search didn't work, it's fine, we just return first block
+                return U64::zero();
+            }
+            current_block = filter_to_block;
+        }
+    }
+
+    pub async fn get_all_blocks_to_process(
+        &self,
+        blob_client: &Arc<dyn BlobClient>,
+        commit_blocks_cache: Option<&Arc<dyn ObjectStore>>,
+        stop_receiver: &watch::Receiver<bool>,
+    ) -> Vec<CommitBlock> {
+        let end_block = L1Fetcher::get_last_l1_block_number(&self.eth_client)
+            .await
+            .unwrap();
+        self.get_blocks_to_process(blob_client, end_block, commit_blocks_cache, stop_receiver)
+            .await
+    }
+
+    pub async fn get_genesis_root_hash(&self) -> H256 {
+        let genesis_block = self.find_block_near_genesis().await;
+        let logs = self
+            .get_logs(
+                genesis_block,
+                genesis_block + self.config.block_step,
+                RollupEventType::Commit,
+            )
+            .await;
+        let tx =
+            L1Fetcher::get_transaction_by_hash(&self.eth_client, logs[0].transaction_hash.unwrap())
```

Comment:
if you don't want to make these methods of L1Fetcher, then consider wrapping eth_client into yet another type instead and making functions like get_transaction_by_hash methods of that new type instead.